src/DSP/Reverter.cpp

```diff
@@ -0,0 +1,281 @@
+/*
+  ZynAddSubFX - a software synthesizer
+
+  Reverter.cpp - Reverse Delay
+  Copyright (C) 2023-2024 Michael Kirchner
+  Author: Michael Kirchner
+
+  This program is free software; you can redistribute it and/or
+  modify it under the teabs_value of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+*/
+
+#include <cassert>
+#include <cstdio>
+#include <cmath>
+#include <cstring>
+
+#include "../Misc/Allocator.h"
+#include "../Misc/Util.h"
+#include "Reverter.h"
+
+namespace zyn {
+
+Reverter::Reverter(Allocator *alloc, float delay_, unsigned int srate, int bufsize, float tRef_, const AbsTime *time_)
+    : syncMode(AUTO), input(nullptr), gain(1.0f), delay(delay_), phase(0.0f), crossfade(0.16f),
+      tRef(tRef_), buffer_offset(0), buffer_counter(0), reverse_index(0.0f), phase_offset_old(0.0f),
+      phase_offset_fade(0.0f), fade_counter(0), mean_abs_value(999.9f), time(time_), memory(*alloc), 
+      samplerate(srate), buffersize(bufsize) 
+{
+
+    // current number of samples to be used for crossfade
+    fading_samples = static_cast<int>(srate * crossfade);
+
+    // maximum number of samples for a reversed segment.
+    max_delay = srate * MAX_REV_DELAY_SECONDS;
+    // Calculate mem_size for reverse delay effect:
+    // - 1 times max_delay for recording
+    // - 1 times max_delay for reverse playing
+    // - 1 times max_delay for phase changing while playing
+    // - Add maximum crossfade duration (1.27s).
+    // - Add 2 extra samples as a safety margin for interpolation and circular buffer wrapping.
+    mem_size = static_cast<int>(ceilf(max_delay * 3.0f)) + static_cast<int>(1.27f * samplerate) + 2;
+
+    input = static_cast<float *>(memory.alloc_mem(mem_size * sizeof(float)));
+    reset();
+
+    setdelay(delay);
+    global_offset = fmodf(tRef, delay);
+    pos_writer = 0;
+    pos_reader = 0;
+    pos_start = 0;
+    reverse_index = 0;
+    state = PLAYING;
+}
+
+Reverter::~Reverter() {
+    memory.dealloc(input);
+}
+
+inline float Reverter::sampleLerp(float *smp, float pos) {
+    int poshi = static_cast<int>(pos);
+    float poslo = pos - static_cast<float>(poshi);
+    return smp[poshi] + poslo * (smp[poshi + 1] - smp[poshi]);
+}
+
+inline float hanningWindow(float x) {
```

Comment:
Based on your comments that removing this eliminated clicks during fading, is this dead code?

Reply:
right